packages/nx/src/native/cache/http_remote_cache.rs

```diff
@@ -0,0 +1,206 @@
+    url: String,
+}
+
+#[napi]
+impl HttpRemoteCache {
+    #[napi(constructor)]
+    pub fn new() -> Self {
+        let mut headers = header::HeaderMap::new();
+        let auth_token = env::var("NX_SELF_HOSTED_REMOTE_CACHE_ACCESS_TOKEN");
+        if let Ok(token) = auth_token {
+            headers.insert(
+                header::AUTHORIZATION,
+                header::HeaderValue::from_str(&format!("Basic {}", token)).unwrap(),
+            );
+        }
+
+        headers.insert(
+            header::CONTENT_TYPE,
+            header::HeaderValue::from_static("application/octet-stream"),
+        );
+
+        HttpRemoteCache {
+            client: ClientBuilder::new()
+                .default_headers(headers)
+                .build()
+                .unwrap(),
+            url: env::var("NX_SELF_HOSTED_REMOTE_CACHE_SERVER")
+                .expect("NX_REMOTE_CACHE_URL must be set"),
+        }
+    }
+
+    #[napi]
+    pub async fn retrieve(
+        &self,
+        hash: String,
+        cache_directory: String,
+    ) -> napi::Result<Option<CachedResult>> {
+        let url: String = format!("{}/v1/cache/{}", self.url, hash);
+        let response = self.client.get(&url).send().await;
+        if let Ok(resp) = response {
+            trace!("HTTP response status: {}", resp.status());
+            if resp.status().is_success() {
+                // response is an application/octet-stream containing a tarball
+                // we need to extract the tarball and return the path to the extracted files
+                Ok(Some(
+                    Self::download_and_extract_from_result(resp, cache_directory.clone()).await?,
+                ))
+            } else {
+                if resp.status().as_u16() == 404 {
+                    trace!("Cache not found for hash '{}'", hash);
+                    return Ok(None); // Return None if the cache was not found
+                }
+                // Handle other HTTP errors
+                Err(anyhow::anyhow!("Failed to retrieve cache: {}", resp.status()).into())
+            }
+        } else {
+            trace!(
+                "Failed to retrieve cache for hash '{}': {}",
+                hash,
+                response
+                    .as_ref()
+                    .err()
+                    .map(|e| e.to_string())
+                    .unwrap_or_default()
+            );
+            Err(anyhow::anyhow!("Failed to retrieve cache: {}", response.unwrap_err()).into())
+        }
+    }
+
+    #[napi]
+    pub async fn store(
+        &self,
+        hash: String,
+        cache_directory: String,
+        terminal_output: String,
+        code: u32,
+    ) -> Result<bool, napi::Error> {
+        // create a tarball in memory from the cache dir
+        let tar_gz: Vec<u8> = Vec::new();
+        let enc = flate2::write::GzEncoder::new(tar_gz, Compression::default());
+        let mut archive = Builder::new(enc);
+        trace!("Created tar file for writing");
```

Comment:
To make these traces more helpful, we can create a span with the hash as the parent. Then when all these logs come out, they'll be labeled with their hash. 

https://docs.rs/tracing/latest/tracing/index.html#spans